### DISCUSSION TITLE
Troubleshooting Assistance

### DISCUSSION THREAD

**Discussion by patrickenfuego** [patrickenfuego/Chapterize-Audiobooks#14] [Q&A]
If you're having problems with the project, please post your questions here.

**Comment by reynold1** [comment]
In PowerShell I ran this command "python .\chapterize_ab.py 'C:\Users\D(JR)x2\Desktop\XXX.mp3'" and got this error message:

Traceback (most recent call last):
  File "C:\Users\D(JR)x2\Desktop\Chapterize-Audiobooks-0.6.0\chapterize_ab.py", line 31, in <module>
    from model.models import (
  File "C:\Users\D(JR)x2\Desktop\Chapterize-Audiobooks-0.6.0\model\models.py", line 116, in <module>
    def get_language_features(lang: str) -> tuple[tuple, tuple] | tuple[None, None]:
TypeError: unsupported operand type(s) for |: 'types.GenericAlias' and 'types.GenericAlias'

**Comment by patrickenfuego** [comment]
Hello. May I ask which version of python you are running? The syntax seen in the error message was introduced in Python 3.10, indicating you may be running an older version.

**Comment by silasrogan** [comment]
Thank you for the great work on this! I have been unable to get past the following error: 
`Timecodes list cannot be empty. Exiting...`

The book series I'm listening to does not say `Chapter X` but rather it says `Part X`. 
Could this be the issue? Thank you!

**Reply by lcta0717** [reply]
I'm having the same issue

**Reply by patrickenfuego** [reply]
I'm very sorry for the late reply on this, I've had a lot going on and haven't been able to maintain the project like I initially planned.

You are correct - `Part` has not yet been implemented. I was in the process of massively updating the project to support features like this, but I've been away so long I need to figure out where I was in the development process when I left off last.

When I initially created the script, I did not consider markers apart from `Chapter` (as I was still really new to audiobooks in general). As I listened to more audiobooks, I realized there are other common separators I did not account for, such as `Part`; there are even audiobooks that skip identifiers altogether and just say `One`, `Two`, etc. 

Before my hiatus, I started building 2 new ways to split books:

- Based on the `Part` word. This became challenging as many books also use this to separate groups of chapters as well. I was testing ways to reliably identify these scenarios and split chapters accordingly.
- Books that don't use separators. In this case, I started writing a fallback silence detection algorithm; I find this less than ideal as it is rarely uniform across audiobooks and often causes erroneous chapter breaks (this was the entire reason I started this project), but it acts as a fallback option if the other options aren't feasible.

Things in my life have calmed down a bit, so I'm hoping to dive back into this project soon. More to come, thank you for your patience.

**Reply by prodigalstatius** [reply]
I ran into an interesting edge case on the Chapter splitting. I ran the script on a theological book that included many references to the Bible, often saying things like "In Romans chapter 3" or "in Genesis chapter 6" or "in the third chapter of Ezekiel". 

I ended up with 185 or so chapters, some only a few seconds long. 

I wondered about finding a way to isolate those chapters (maybe adding some logic that excluded a list of words, in this case the names of the books of the Bible" either before or after each word of chapter; though that may not be perfect, if a chapter ends with one of those words, for example).

**Reply by patrickenfuego** [reply]
@prodigalstatius this is an excellent example of how to expand the "excluded phrases" list; this is essentially a list of words (or "phrases", if you will) that, when grouped together, the script detects via pattern matching and ignores during chapter splitting.

If you don't mind and have the time, please compile a list of all (or as many as you can) of these phrases that triggered chapter breaks and submit them in a GitHub issue. From there, I can take them, add them to this list, and it should reduce the number of erroneous chapter breaks. This would ultimately help everyone using this script in the future!

**Reply by prodigalstatius** [reply]
I am traveling at the moment but I may be able to do that this weekend. 

This is such a great project. I've used it a few times. 

I do have a book I want to chapterize that is going to prove a challenge, though. Grimms Fairy Tales are all individual stories but none start with Chapter at the beginning.

**Reply by lcta0717** [reply]
> I ran into an interesting edge case on the Chapter splitting. I ran the script on a theological book that included many references to the Bible, often saying things like "In Romans chapter 3" or "in Genesis chapter 6" or "in the third chapter of Ezekiel".
> 
> I ended up with 185 or so chapters, some only a few seconds long.
> 
> I wondered about finding a way to isolate those chapters (maybe adding some logic that excluded a list of words, in this case the names of the books of the Bible" either before or after each word of chapter; though that may not be perfect, if a chapter ends with one of those words, for example).

This is why I love both language and programming! It’s so complex.
For instance, the code already has “Matthew Chapter” on the avoid list. However, it wouldn’t make sense to put all 66 books on the avoid list, as it’s really common to simply refer to the chapter and verse if you’ve already mentioned the book (I.e. “In John chapter three … and in chapter four, Jesus goes down to Samaria…”). This makes it anything but easy to figure out what would actually be a chapter division.

I previously found in the code where the list of division keywords were, and simply added “tale” to my own local copy (I’m trying to split Beatrix Potter for my son). However, the second book (Squirrel Nutkin) starts off with the line, “This is a tale about a tail.” The AI put a division on both. It’s amazing to see how important context is!

**Reply by lcta0717** [reply]
I wonder if you would be able to expose the chapter divisions after the AI finds them, such as in a file? I’m not sure if they’re shared as timestamps or something else, but if we could have access to the list before the program splits the mp3, we could manually adjust for all the erroneous (and really difficult) divisions…
